### PR TITLE
ARM template component configuration integration

### LINKED ISSUE #944: Create ARM template for Daemon VM
@imedina @j-coll @wbari can you provide some more details as to what this needs? 

As far as I know:

- [x] Open CGA Daemon in Docker - just needs OpenCGA and Java - no tomcat?
- [ ] Tooling - what?
- [x] Spec/sizing - what cpu/ram/disks etc?

### PR DESCRIPTION
**Pull request by marrobi**
DO NOT MERGE

https://github.com/opencb/opencga/pull/970 should be merged first.

Closes #944 .

@jjcollinge to pick this up.

### CODE REVIEW

File: opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        "vnetId": "[parameters('virtualNetworkId')]",
+        "subnetRef": "[concat(variables('vnetId'), '/subnets/', parameters('subnetName'))]",
+        "cloud-init": [
+
+            "#cloud-config",
+            "package_upgrade: true",
+            "package_reboot_if_required: true",
+            "",
+            "packages:",
+            "- docker.io",
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: Nit: 'docker.io' is the Ubuntu repository version of Docker so will lag behind the official repo from docker.

Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: Believe this will give 17.x while official docker repo is now 18.x

**Comment by marrobi** [comment]
It's changed recently. Good explanation... https://www.collabora.com/news-and-blog/blog/2018/07/04/docker-io-debian-package-back-to-life/. It was just easier, happy to change.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

File: opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json (hunk 1 above)
Comment (marrobi) [reply opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: It's changed recently. Good explanation... https://www.collabora.com/news-and-blog/blog/2018/07/04/docker-io-debian-package-back-to-life/. It was just easier, happy to change.

Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: I'd use something like 
```
runcmd:
- apt-get update
- curl -fsSL https://get.docker.com/ | sh
- curl -fsSL https://get.docker.com/gpg | sudo apt-key add -
```

cc @jjcollinge as he's picking this up

Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: as we don't know how up-to-date the deb package will be kept

**Comment by jjcollinge** [comment]
once #978 has been merged, I'll merge #981 into here and then hook up the Solr details into the init script. I'll also pick up the loadbalancer IP changes from #980  and inject those as rest/grpc hosts. Mongo is externally managed at the moment, so those details can just be dropped in as top level parameters. Then I should be able to start ironing out any creases in the e2e deployment.

File: opencga-app/app/scripts/azure/arm/azuredeploy.json [hunk 2]
```diff
@@ -229,6 +273,82 @@
...
+                        "value":  "[reference('hdinsight').outputs.clusterSshPassword.value]"
+                    },
+                    "daemonContainerImage": {
+                        "value": "[parameters('daemonContainerImage')]"
+                    },
+                    "initContainerImage": {
+                        "value": "[parameters('initContainerImage')]"
+                    },
+                    "mountArgs": {
+                        "value": "[concat('azurefiles ', reference('azurefiles').outputs.accountName.value, ',', reference('azurefiles').outputs.shareName.value, ',' , reference('azurefiles').outputs.accountKey.value)]"
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/azure/arm/azuredeploy.json]: To pick up and use the switch from #980 you can change this line to switch between `avere` and `azurefiles` like this:

```suggestion
                         "value": "[if(equals(parameters('nfsStorageOption'), 'AzureFiles'), concat('azurefiles ', reference('azurefiles').outputs.accountName.value, ',', reference('azurefiles').outputs.shareName.value, ',' , reference('azurefiles').outputs.accountKey.value),concat('avere ', reference('avere').outputs.VSERVER_IPS.value))]"
```

File: opencga-app/app/scripts/azure/arm/webservers/azuredeploy.json [hunk 3]
```diff
@@ -115,8 +115,8 @@
             "- curl -fsSL https://get.docker.com/ | sh",
             "- curl -fsSL https://get.docker.com/gpg | sudo apt-key add -",
             "[concat('- curl -fsSL ', parameters('_artifactsLocation'), '/avere/mount.py', parameters('_artifactsLocationSasToken'), ' | python3 - ', parameters('mountArgs'))]",
-            "[concat('- docker run -d -p 8080:80 -p 8443:443 --restart=always -v /media/primarynfs:/nfs ', parameters('tomcatDockerImage'))]",
-            "[concat('- docker run -d -p 80:80 -p 443:443 --restart=always -v /media/primarynfs:/nfs ', parameters('ivaDockerImage'))]",
+            "[concat('- docker run --user 1005 -d -p 8080:80 -p 8443:443 --restart=always --mount type=bind,src=/media/primarynfs/conf,dst=/opt/opencga/conf,readonly --mount type=bind,src=/media/primarynfs/sessions,dst=/opt/opencga/sessions ', parameters('tomcatDockerImage'))]",
```
Comment (lawrencegripper) [inline opencga-app/app/scripts/azure/arm/webservers/azuredeploy.json]: Thanks for fixing this up!

File: opencga-app/app/scripts/azure/arm/azuredeploy.json (hunk 2 above)
Comment (jjcollinge) [reply opencga-app/app/scripts/azure/arm/azuredeploy.json]: thanks!

File: opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        },
+        "catalogSecretKey": {
+            "type": "securestring"
+        }
+    },
+    "variables": {
+        "vnetId": "[parameters('virtualNetworkId')]",
+        "subnetRef": "[concat(variables('vnetId'), '/subnets/', parameters('subnetName'))]",
+        "restHost": "[concat('https://', parameters('loadBalancerIp'), ':8443')]",
+        "grpcHost": "[concat('https://', parameters('loadBalancerIp'), ':9999')]",
```
Comment (jjcollinge) [inline opencga-app/app/scripts/azure/arm/daemonvm/azuredeploy.json]: Get port from Load Balancing rules?

**Pull request merged**